freqtrade/exchange/exchange.py

```diff
@@ -1525,6 +1581,61 @@ def get_historic_trades(self, pair: str,
             self._async_get_trade_history(pair=pair, since=since,
                                           until=until, from_id=from_id))
 
+    @retrier
+    def fill_leverage_brackets(self):
+        """
+            #TODO-lev: Should maybe be renamed, leverage_brackets might not be accurate for kraken
+            Assigns property _leverage_brackets to a dictionary of information about the leverage
+            allowed on each pair
+        """
+        raise OperationalException(
+            f"{self.name.capitalize()}.fill_leverage_brackets has not been implemented.")
+
+    def get_max_leverage(self, pair: Optional[str], nominal_value: Optional[float]) -> float:
+        """
+            Returns the maximum leverage that a pair can be traded at
+            :param pair: The base/quote currency pair being traded
+            :nominal_value: The total value of the trade in quote currency (collateral + debt)
+        """
+        raise OperationalException(
+            f"{self.name.capitalize()}.get_max_leverage has not been implemented.")
+
+    @retrier
+    def set_leverage(self, leverage: float, pair: Optional[str]):
+        """
+            Set's the leverage before making a trade, in order to not
+            have the same leverage on every trade
+        """
+        try:
+            self._api.set_leverage(symbol=pair, leverage=leverage)
+        except ccxt.DDoSProtection as e:
+            raise DDosProtection(e) from e
+        except (ccxt.NetworkError, ccxt.ExchangeError) as e:
+            raise TemporaryError(
+                f'Could not set leverage due to {e.__class__.__name__}. Message: {e}') from e
+        except ccxt.BaseError as e:
+            raise OperationalException(e) from e
+
+    @retrier
+    def set_margin_mode(self, pair: str, collateral: Collateral, params: dict = {}):
+        '''
+            Set's the margin mode on the exchange to cross or isolated for a specific pair
+            :param symbol: base/quote currency pair (e.g. "ADA/USDT")
+        '''
+        if not self.exchange_has("setMarginMode"):
+            # Some exchanges only support one collateral type
+            return
+
+        try:
+            self._api.set_margin_mode(pair, collateral.value, params)
+        except ccxt.DDoSProtection as e:
+            raise DDosProtection(e) from e
+        except (ccxt.NetworkError, ccxt.ExchangeError) as e:
+            raise TemporaryError(
+                f'Could not set margin mode due to {e.__class__.__name__}. Message: {e}') from e
+        except ccxt.BaseError as e:
+            raise OperationalException(e) from e
```

Comment:
This is binance only (no other exchange supports "setMarginMode" at this moment.

I'd therefore move this to the binance class, so it's clear that it's "binance-specific" code.

Reply:
I think my response is the same as under `set_leverage`. Currently ccxt only has this for Binance, but they're goal is to get it for each exchange that makes sense which would include ByBit, and any exchange that uses both isolated and cross

If I move it to Binance, there's the pro of making it clear that it's a method which only works for binance right now, but if it gets implemented for another exchange, then we'll just have to move it back for it to work for that exchange

Reply:
Works for me :+1: